### DISCUSSION TITLE
What will be the best method to access the services exposed by TopLevel in MVVM?

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#11419] [Q&A]
Avalonia Team is progressively moving some global services like Clipboard , StorageProvider and etc.. #11381 in TopLevel.

This choice causes several problems to access these services in MVVM. I would like to discuss with you what would be the best choice.

I'm currently using slightly modified wieslawsoltes hint [code](https://github.com/AvaloniaUI/Avalonia/discussions/11093#discussioncomment-5686977).

```csharp
public static TopLevel? GetTopLevel(this Application app)
{
    if (app.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
    {
        return desktop.MainWindow;
    }
    if (app.ApplicationLifetime is ISingleViewApplicationLifetime viewApp)
    {
        var visualRoot = viewApp.MainView?.GetVisualRoot();
        return visualRoot as TopLevel;
    }
    // Where app has not ApplicationLifetime try to get TopLevel from current focused element
    return TopLevel.GetTopLevel(FocusManager.Instance?.Current as Visual); 
}
```

When PR #11407 is unity, the statment

```csharp
TopLevel.GetTopLevel(FocusManager.Instance?.Current as Visual)
````
it will no longer be valid.

another possible solution could be this:

#### ItemModelView.xaml.cs

```csharp
   protected override void OnLoaded()
   {
       base.OnLoaded();
       if (this.DataContext is ViewModels.ItemModel model)
       {
            model.TopLevel = TopLevel.GetTopLevel(this);
       }
  }
```

I don't like this solution because:
- You need to add the above code to each view
- It could result in a memory leak

Do you have any other idea?

**Comment by kekekeks** [comment]
I'd suggest to use the same idea as frameworks like RxUI/Caliburn: those have special base classes for views and view models, which are aware of each other. So a view model can access relevant services once it's attached to something on screen.

Also note, that in general it's a bad idea to use MainWindow global instance for file dialogs since it can have undesired results when a modal dialog is onscreen.

**Reply by workgroupengineering** [reply]
Thanks for the reply, I'll take a look at it. I leave the discussion open in case there is any other idea suggestion.

**Reply by timunie** [reply]
Maybe part of this answer can be useful here as well: https://github.com/AvaloniaUI/Avalonia/discussions/11466#discussioncomment-5964123 

especially: 

> Use a message bus, e.g.:
> - https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/mvvm-messenger-and-view-services-in-mvvm
> - https://www.reactiveui.net/docs/handbook/message-bus/
> - https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/messenger
> 
> Use Interaction
> - https://www.reactiveui.net/docs/handbook/interactions/
> - https://docs.avaloniaui.net/tutorials/music-store-app/opening-a-dialog
